libfreerdp/codec/progressive.c

```diff
@@ -392,38 +455,103 @@ static PROGRESSIVE_SURFACE_CONTEXT* progressive_surface_context_new(UINT16 surfa
 		free(surface);
 		return NULL;
 	}
+	for (x = 0; x < surface->gridSize; x++)
+	{
+		if (!progressive_tile_allocate(&surface->tiles[x]))
+		{
+			progressive_surface_context_free(surface);
+			return NULL;
+		}
+	}
 
 	return surface;
 }
 
-static void progressive_surface_context_free(PROGRESSIVE_SURFACE_CONTEXT* surface)
+static BOOL progressive_surface_tile_replace(PROGRESSIVE_SURFACE_CONTEXT* surface,
+                                             PROGRESSIVE_BLOCK_REGION* region,
+                                             const RFX_PROGRESSIVE_TILE* tile, BOOL upgrade)
 {
-	UINT32 index;
-	RFX_PROGRESSIVE_TILE* tile;
+	RFX_PROGRESSIVE_TILE* t;
 
-	for (index = 0; index < surface->gridSize; index++)
+	size_t zIdx;
+	if (!surface || !tile)
+		return FALSE;
+
+	zIdx = (tile->yIdx * surface->gridWidth) + tile->xIdx;
+
+	if (zIdx >= surface->gridSize)
 	{
-		tile = &(surface->tiles[index]);
+		WLog_ERR(TAG, "Invalid zIndex %" PRIuz, zIdx);
+		return FALSE;
+	}
 
-		if (tile->data)
-			_aligned_free(tile->data);
+	t = &surface->tiles[zIdx];
 
-		if (tile->sign)
-			_aligned_free(tile->sign);
+	if (upgrade)
+	{
```

Comment:
I can't see any difference between the `if` and the `else` statement, am I missing something ?

Reply:
@hardening some of the struct members are different. was too lazy to extract them from the if/else struct